instrumentation/action_pack/lib/opentelemetry/instrumentation/action_pack/patches/action_controller/live.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module ActionPack
+      module Patches
+        module ActionController
+          # Module to append to ActionController::Live for instrumentation
+          module Live
+            def process_action(*)
+              current_context = OpenTelemetry::Context.current
+
+              # Unset thread local to avoid modifying stack array shared with parent thread
+              Thread.current[:__opentelemetry_context_storage__] = nil
```

Comment:
This is bad. We shouldn't have to reach into the API internals in this way. This happens because `OpenTelemetry::Context.clear` is implemented as `stack.clear`, which empties the backing array instead of replacing it with a new empty instance.

I think it'd be worthwhile to change the implementation of `OpenTelemetry::Context.clear` to be `Thread.current[STACK_KEY] = []`.

On the other hand, I think there is a race condition when fetching the current context. Specifically, since the thread-locals are copied, the backing array for the context stack is shared with the parent thread, and if that parent thread modifies its context before _this_ thread retrieves it, the context used in this thread will be incorrect.

Side note: the tracing library we built at Shopify before OpenTelemetry Ruby patched `ActionController::Live` in a very similar way, other than unsetting the thread local. It worked there because the thread-local was a `Span` instead of an array, and we managed the "stack" as a linked list of spans.

I think it might be preferable to explicitly propagate context in some way, but I'm not familiar enough with `ActionController::Live` to know how to accomplish that.

Reply:
Thanks @fbogsany. I will open a PR in `opentelemetry-ruby` to propose your suggested change to `Context.clear`. I'll also probably close this PR and reopen a new one with a different approach, using explicit propagation like you suggest.

Reply:
I tried a linked list implementation of the context stack in https://github.com/open-telemetry/opentelemetry-ruby/pull/1597 but the performance impact was significant (~9% with the interpreter, as much as 24% with yjit).

Reply:
https://github.com/open-telemetry/opentelemetry-ruby/pull/1598

Reply:
https://github.com/open-telemetry/opentelemetry-ruby/pull/1598 is merged. Once we cut a new release of the API, we can use that as a minimum version here.

Reply:
@thomasmarshall  - A new version of the API has been released! Could you incorporate it into this PR?